% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/binom.test.QTL.R
\name{binom.test.QTL}
\alias{binom.test.QTL}
\title{Binomial test of SNP QTL}
\usage{
binom.test.QTL(gff, qtl_table, s, s_v, genome_length, chr_col, pos_col, snp_col,
  alternative = c("two.sided", "less", "greater"))
}
\arguments{
\item{gff}{Path to the global QTL file including multiple phenotypes}

\item{qtl_table}{Path to the SNP QTL file for a given phenotype, csv format}

\item{s}{is boolean for subsetting of breeds}

\item{s_v}{is a vector with the breeds to be subsetted on}

\item{genome_length}{is the length in base pairs of the species analysed}

\item{chr_col}{chromosome name column, must be in gff format (Chr.nr)}

\item{pos_col}{SNP position column}

\item{snp_col}{SNP id column}

\item{alternative}{binomial alternative test hypothesis}
}
\value{
dataframe of frequency overlap and binomial probability
}
\description{
This function runs the binomial test for each phenotype in global QTL dataframe against
SNP QTL. It can be used to support QTLs found in a study, and find novel interactions between different traits and the trait used in the study.
}
